### PR TITLE
Add websocket support to RPC

### PR DESCRIPTION
**Pull request by TheQuantumPhysicist**

### CODE REVIEW

File: node/src/options.rs [hunk 1]
```diff
@@ -92,9 +92,21 @@ pub struct RunOptions {
...
+    #[clap(long)]
+    pub http_rpc_enabled: Option<bool>,
+
+    /// Address to bind websocket RPC to.
+    #[clap(long, value_name = "ADDR")]
+    pub ws_rpc_addr: Option<SocketAddr>,
+
+    /// Enable/Disable websocket RPC.
+    #[clap(long)]
+    pub ws_rpc_enabled: Option<bool>,
```
Comment (iljakuklic) [inline node/src/options.rs:109]: Here's not very well thought out suggestion. Maybe the combination of the `SocketAddr` and `bool` could be handled by having a special value for disabling the address, like `--ws-rpc-addr=none` or `--ws-rpc-addr=disable`?

Using a custom type like
```rust
enum SocketAddrConfig { Disabled, Enabled(SorcketAddr) }
impl FormStr for SocketAddrConfig { /* handle "none" here */ }
```
the `ws_rpc_addr` and `http_rpc_addr` could be `Option<SocketAddrConfig>`.

I don't have a problem with the current approach for now, the configuration handling needs a bit more thought anyway so can be done separately.

File: node/src/runner.rs [hunk 2]
```diff
@@ -104,16 +104,19 @@ pub async fn initialize(
...
+            "rpc",
+            rpc::Builder::new(node_config.rpc)
+                .register(chainstate.clone().into_rpc())
+                .register(mempool.into_rpc())
+                .register(NodeRpc::new(manager.make_shutdown_trigger()).into_rpc())
+                .register(p2p.clone().into_rpc())
+                .build()
+                .await?,
+        );
+    }
```
Comment (iljakuklic) [inline node/src/runner.rs:119]: Would it make sense for `rpc-http` and `rpc-ws` to be separate subsystems and only start the ones that are enabled? The `rpc` crate now has way too many pattern matches on `Option` to my liking.

After some refactoring in the `rpc` module, this bit could look something like:

```rust
let rpc_builder = rpc::builder::new().register(/*...*/).register(/*...*/);
if node_config.rpc.http_enabled() {
    manager.add_subsystem("rpc-http", rpc_builder.build_http(node_config.rpc.http_addr()).await);
}
if node_config.rpc.ws_enabled() {
    manager.add_subsystem("rpc-ws", rpc_builder.build_ws(node_config.rpc.ws_addr()).await);
}
```

**Review by iljakuklic — COMMENTED** [review]
Adding some initial comments

File: node/src/options.rs (hunk 1 above)
Comment (TheQuantumPhysicist) [reply node/src/options.rs:109]: I don't want to deal with this until the configuration issue is solved. It's basically beating a dead horse at this point.

File: node/src/runner.rs (hunk 2 above)
Comment (TheQuantumPhysicist) [reply node/src/runner.rs:119]: They both use the same function objects, so we'll have to duplicate the handlers. Not really the best choice.

**Review by azarovh — APPROVED** [review]

File: node/src/options.rs (hunk 1 above)
Comment (TheQuantumPhysicist) [reply node/src/options.rs:109]: To be clear, I'm talking about #446 . This has wasted enough time in this PR.

**Comment by TheQuantumPhysicist** [comment]
> I'm fine with merging this. However, I'd say having a separate sybsystem for each server (i.e. `rpc-http` and `rpc-ws`) deserves some consideration

I'll have to work on websocket much more later. Let's solve one problem at a time. I'm still not sure we need this at this point.

**Review by stanislav-tkach — APPROVED** [review]

**Pull request merged**